linear-algebra/src/main/scala/scalan/linalgebra/Matrices.scala

```diff
@@ -89,112 +101,234 @@ trait Matrices extends Vectors { self: MatricesDsl =>
+          }
+          CompoundMatrix(rowsNew, numColumnsB)
+        case CompoundMatrix(rowsB, numColumnsB) =>
+          val (is, vs) = (replicatedRow.nonZeroIndices, replicatedRow.nonZeroValues)
+          val res = CompoundMatrix((vs zip rowsB(is)).map { case Pair(a, vB) => vB *^ a }, numColumnsB).reduceByColumns
+          // TODO: find a proper way to carry over type of vector (Sparse or Dense)
+          ReplicatedMatrix(res, numRows)
+        case ReplicatedMatrix(rowsB, numColumnsB) =>
+          val vs = matrix.replicatedRow
+          val res = CompoundMatrix((vs zip rowsB(is)).map { case Pair(a, vB) => vB *^ a }, numColumnsB).reduceByColumns
+
+        case ConstMatrix(value, width, _) =>
+          val rowsConstant = (reduceByRows *^ value).items.map(v => ConstVector(v, width))
+          CompoundMatrix(rowsConstant, width)
+        case DiagonalMatrix(diagonalValues) =>
+          val diagonalVector = DenseVector(diagonalValues)
+          CompoundMatrix(rows.map(row => row *^ diagonalVector), numColumns)
+        case ConstDiagonalMatrix(diagonalValue, _) =>
+          CompoundMatrix(rows.map(row => row *^ diagonalValue), numColumns)
+        // TODO: check case _
+        case _ => !!!("matcher for @matrix argument in CompoundMatrix.*(matrix: Matr[T]) is not specified.")
+      }
+    }
+
+    @OverloadId("matrix")
+    def +^^(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
+      def res = CompoundMatrix((rows zip matrix.rows).map { case Pair(v, s) => v +^ s }, numColumns)
+      matrix match {
+        case CompoundMatrix(_, _) => res
+        case ConstMatrix(_, _, _) => res
+        case DiagonalMatrix(_) => res
+        case ConstDiagonalMatrix(_, _) => res
+        case _ => matrix +^^ self
+      }
+    }
+
+    @OverloadId("matrix")
+    def *^^(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
+      def res = CompoundMatrix((rows zip matrix.rows).map { case Pair(v, s) => v *^ s }, numColumns)
+      matrix match {
+        case CompoundMatrix(_, _) => res
+        case ConstMatrix(_, _, _) => res
+        case DiagonalMatrix(diagonal) =>
+          DiagonalMatrix((diagonalValues zip diagonal).map { case Pair(v, d) => v * d })
+        case ConstDiagonalMatrix(diagonalValue, width) =>
+          DiagonalMatrix(diagonalValues.map(v => v * diagonalValue))
+        case _ => matrix *^^ self
+      }
+    }
+
+    def average(implicit f: Fractional[T], m: RepMonoid[T]): Rep[T] = {
+      val items = rows.flatMap(v => v.nonZeroValues)
+      items.reduce / (numRows * numColumns).to[T]
+    }
+  }
+*/
   abstract class CompoundMatrix[T](val rows: Rep[Collection[Vector[T]]], val numColumns: Rep[Int])
                                   (implicit val eT: Elem[T]) extends Matrix[T] {
 
-    def companion = CompoundMatrix
     def columns: Rep[Collection[Vector[T]]] = {
+      // TODO: inefficient
       Collection(SArray.tabulate(numColumns) { j => DenseVector(rows.map(_(j)))})
```

Comment:
It is not only in priorities, it should be implemented and merged first. And only then other dependent features should be implemented.